### PR TITLE
Add internal package to libandroid-navigation

### PR DESCRIPTION
**Pull request by Guardiola31337**
- Adds `internal` package to `libandroid-navigation` and places the APIs which are designed only for local usage

As this is going to be a _big_ refactor the plan here is to open _small_ PRs against `libandroid-navigation` branch moving stuff incrementally and making the review easier. Even trying to do so this first PR included changes in 40 files - this is what happens when a bunch of things depend on each other 😬 So noting that breaking this in smaller chunks will force us to include some `TODO`s that we will need to 👀 in the following PRs, until we're happy with the whole refactor so we will be able to merge.

This first PR also should serve as a team discussion channel on how to tackle this effort so that we can apply the same approach in `libandroid-navigation-ui` module later.

cc @akitchen

### CODE REVIEW

**Comment by Guardiola31337** [comment]
@danesfeder 
> I'm wondering, do these `public` classes in the `internal` package behavior any differently than `public` classes outside of the `internal` package?  Or developers will just know that because they are in the `internal` package, they should not trust to use them in their own project.

Nah, you're right, these classes behave as any other `public` class outside of the `internal` package but for being under the `internal` package developers should know that are internal-only implementation details and shouldn't depend on them because they might change between releases.

One thing that we could explore though is how to make the `internal` classes less discoverable from the IDE:
- Maybe using the `@hide` annotation? I know that [`metalava`](https://android.googlesource.com/platform/tools/metalava/) does some kind of magic around these lines 👀 
> more importantly to hide code intended to be implementation only, driven by javadoc comments like `@hide`, `@$doconly`, `@removed`, etc, as well as various annotations.
- Naming classes without the most "common" prefixes
- ...

What do you think? Anything else that comes to mind?

In any case, this is something that we can address in parallel / follow up after this PR lands.

**Comment by Guardiola31337** [comment]
@devotaaabel 
> I'm wondering about the public classes. I'm guessing those are classes that are still currently needed by classes not in the `internal` package? Will they remain public? Should there be `todo`s for making them package protected?

I see what you mean but this is the whole point of this effort. By placing the internal-only implementation classes under the `internal` package we're making implicitly clear to developers that they shouldn't depend on them because they could change between releases. There will be obviously some classes that will change from being package-private to `public` but that shouldn't be a problem. IMO "structure" >>> "modifiers visibility" especially in this case in which developers are aware of how `internal` packages "work" / are meant to. This is basically a constraint from the language itself and the modifiers that are available. Another option would be having no packages at all and rely on the modifiers, but I think this would become unmanageable / a mess as we keep adding new functionalities / classes. Ultimately, this is a common practice in SDK / library development.

That being said we should take advantage of this overhaul and try to hide the visibility of the APIs exposed (both internal and public classes) and expose only which is strictly necessary avoiding confusion to developers.

> Otherwise, it's unfortunate to have to make more classes public, even if we are designating them `internal`.

Why do you think it's unfortunate? Could you clarify your point here? 🤔
Want to note that we won't need to strictly document these new "public" classes (if that's what worries you).

> Maybe adding some more comments in the public `internal` classes would be good, to make sure we are very explicit about the intention.

This could go under the things that we could implement to make these internal-only APIs less discoverable. Although as I pointed out I wouldn't rush too much about this right now, mainly because most of the clients are aware of the `internal` package purpose and if we notice that developers are messing up we can always work on this later.

**Comment by Guardiola31337** [comment]
@danesfeder 
> as a start, maybe we can add [`@hide`](https://stackoverflow.com/questions/17035271/what-does-hide-mean-in-the-android-source-code) to all `public` classes in the `internal` package?

Yeah we could definitely do that but I think `internal` package is good enough for now - devs know (should know 😅) that classes under `internal` package are designed only for local usage.

In any case, thanks for bringing this up because another thing that we need to take care of here is to "exclude" the "internal" Javadoc from being published. We can follow up on this later though, as this is going to land in the `libandroid-navigation-internal` feature branch.

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationEventDispatcher.java [hunk 1]
```diff
@@ -19,7 +20,7 @@
 
 import timber.log.Timber;
 
-class NavigationEventDispatcher {
+public class NavigationEventDispatcher {
```
Comment (devotaaabel) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationEventDispatcher.java]: final?

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationRouteCallback.java [hunk 2]
```diff
@@ -8,13 +8,13 @@
 import retrofit2.Callback;
 import retrofit2.Response;
 
-class NavigationRouteCallback implements Callback<DirectionsResponse> {
+public class NavigationRouteCallback implements Callback<DirectionsResponse> {
```
Comment (devotaaabel) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationRouteCallback.java]: final?

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationRouteEventListener.java [hunk 3]
```diff
@@ -1,13 +1,13 @@
-package com.mapbox.services.android.navigation.v5.navigation;
+package com.mapbox.services.android.navigation.v5.internal.navigation;
 
 import okhttp3.Call;
 import okhttp3.EventListener;
 
-class NavigationRouteEventListener extends EventListener {
+public class NavigationRouteEventListener extends EventListener {
```
Comment (devotaaabel) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationRouteEventListener.java]: final?

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationTelemetry.java [hunk 4]
```diff
@@ -36,7 +39,7 @@
 import java.util.TimerTask;
 import java.util.concurrent.TimeUnit;
 
-class NavigationTelemetry implements NavigationMetricListener {
+public class NavigationTelemetry implements NavigationMetricListener {
```
Comment (devotaaabel) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationTelemetry.java]: final?

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/SdkVersionChecker.java [hunk 5]
```diff
@@ -1,4 +1,4 @@
-package com.mapbox.services.android.navigation.v5.navigation;
+package com.mapbox.services.android.navigation.v5.internal.navigation;
 
 public class SdkVersionChecker {
```
Comment (devotaaabel) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/SdkVersionChecker.java]: final?

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/MapboxNavigator.java [hunk 6]
```diff
@@ -13,7 +13,8 @@
 
 import java.util.Date;
 
-class MapboxNavigator {
+// TODO internal
+public class MapboxNavigator {
```
Comment (devotaaabel) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/MapboxNavigator.java]: final? Also remove 'TODO'

Comment (Guardiola31337) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/MapboxNavigator.java]: Nah, that's a `TODO` that we need to check before merging `libandroid-navigation-internal` into `master`: Could `MapboxNavigator` live in the `internal` package?

**Comment by Guardiola31337** [comment]
OK, going ahead and merging here, thanks for the great feedback @danesfeder and @devotaaabel as discussed let's keep it coming and iterate over `libandroid-navigation-internal` branch.

**Pull request merged**